### PR TITLE
first draft of iam-token script

### LINKED ISSUE #10940: Add support for TOKEN based authentication
**Impact of the new feature**
We should provide support for token based authentication with external services. We gradually move all cmsweb service to new authentication mechanism and x509 authentication will be phase out.

**Is your feature request related to a problem? Please describe.**
The x509 authentication will be dropped in a near by future and all clients should adopt token based authentication.

**Describe the solution you'd like**
adjust HTTP request handlers and any other pieces within WMCore services to support TOKEN based authentication;

**Describe alternatives you've considered**
None
**Additional context**

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #10940 

#### Status
ready

#### Description
New script to obtain IAM token from client's credentials (which should be obtained separately from IAM provider)

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
#10939 

#### External dependencies / deployment changes
`curl` and `jq` tools

#### Additional notes
This script can be used on remote WMAgents to obtain fresh IAM tokens and should be run periodically, as cronjob. We only need to clarify the periodicity of obtaining the token, but I would assume once per day would be sufficient.

### CODE REVIEW

**Comment by vkuznet** [comment]
I tested the script with newly created IAM credentials. Here are the steps to obtain IAM credentials:
- visit https://cms-auth.web.cern.ch/
- login via your X509 or CERN SSO, if necessary you'll need to register
- then, click on `Self-service client registration`
- provide proper fields in the main tab, these includes:
  - Client-Name, e.g. `Test-WMAgent`
  - Redirect URIs, e.g. `https://cmsweb.cern.ch`
  - Contacts, e.g. you may add some e-group which will handle the registration and access to tokens, for that I suggest to created admin group (we do follow that rule in CMSWEB, Monitoring groups) which will have only few members, like L2s or main developers. This group will receive emails or can be used to access to token afterwards
- go to `Access` tab and check all scopes your token will need, I chose compute.read, etc. which may be relevant for WMAgent
- THE MOST IMPORTANT: for `Grant Types` choose `client credentials` and for `Response Types` uncheck everything
- finally, return to main tab and click on Save. This will present you with client's credentials, configuration URI and registration access token. Save all of them in safe place (DO NOT SHARE). The client_id and client_secret will be used by this script and you'll need them to put in place on WMA node. While, registration token will be used if you'll later need to change something in your token configuration.

After you save your `client_id` and `client_secret` you may run the provided script as following:
```
IAM_CLIENT_ID=./secrets/client_id IAM_CLIENT_SECRET=./secrets/client_secret bin/iam-token.sh
``` 
or, put your client_id and client_secret in different location. The script will generate token in /tmp/token which later you can use for access to OAuth protected sites and/or in your workflows.

**Comment by vkuznet** [comment]
@amaltaro , @todor-ivanov how about to review this PR, I submitted it 20 days ago and so far I didn't hear from you anything about it.

**Review by todor-ivanov — COMMENTED** [review]
Thanks @vkuznet 

I did make two comments inline. You may want to take a look. Other than that the code looks like a working one and in a good shape.

**Comment by vkuznet** [comment]
@todor-ivanov I resolved your comments. please review again.

**Review by todor-ivanov — APPROVED** [review]
Thanks @vkuznet  for the prompt fix. 
Things look perfect to me.

**Comment by vkuznet** [comment]
@amaltaro , could you please review it, the PR is pending for a long time.

File: bin/iam-token.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+# the iam-token.sh scripts generate new IAM token
```
Comment (amaltaro) [inline bin/iam-token.sh:2]: Valentin, how about renaming it to `create-iam-token.sh`?

File: bin/iam-token.sh [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# it relies on the following environment variables
+# IAM_CLIENT_ID client id value
+# IAM_CLIENT_SECRET client secret value
+# IAM_TOKEN output file name to store obtained IAM token
```
Comment (amaltaro) [inline bin/iam-token.sh:6]: Maybe we should rename it from `IAM_TOKEN` to `IAM_TOKEN_FILENAME`?

File: bin/iam-token.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# https://github.com/dmwm/WMCore/pull/11093#issuecomment-1098131010
+
+# check if curl exist on a system
+if ! command -v curl &> /dev/null
+then
+    echo "curl could not be found, please install it on your system"
+    exit
+fi
+# check if jq exist on a system
+if ! command -v jq &> /dev/null
```
Comment (amaltaro) [inline bin/iam-token.sh:17]: I checked lxplus and one CERN agent and I confirm this JSON tool exists. Still have to check the FNAL node.
Nonetheless, can you please update the initial PR description mentioning that this `jq` package is required in the node running a token-based WMCore service?

File: bin/iam-token.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+# the iam-token.sh scripts generate new IAM token
+# it relies on the following environment variables
+# IAM_CLIENT_ID client id value
```
Comment (amaltaro) [inline bin/iam-token.sh:4]: Can you please clarify whether these variables are supposed to contain the final value? Or are they suppose to contain a file path which will then contain the value? I guess it's the latter from the code, but we better make it explicit here.

File: bin/iam-token.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    export client_secret=`cat $IAM_CLIENT_SECRET`
+else
+    echo "unable to locate client_secret file, please either setup IAM_CLIENT_secret to point to your client_secret file name"
+    exit
+fi
+#echo "use client_secret=$client_secret"
+
+# obtain new token using client credentials
+if [ -n "IAM_TOKEN" ]; then
+curl -s -k -d grant_type=client_credentials \
```
Comment (amaltaro) [inline bin/iam-token.sh:45]: Can you please explain what these options are? The `client_credentials` value is it really a value or supposed to be a bash variable? Regarding the scope="profile", will it be always like that as well? Maybe we should define constant variables up at the top of this script?

Can you please also indent these two lines within the `if` block?

File: bin/iam-token.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        -d scope="profile" \
+        -u ${client_id}:${client_secret} \
+        https://cms-auth.web.cern.ch/token | jq -r '.access_token' > $IAM_TOKEN
+echo "New IAM token generated and can be found at $IAM_TOKEN"
+else
+    echo "Please etup IAM_TOKEN environment variable pointing to a file name where token will be written"
```
Comment (amaltaro) [inline bin/iam-token.sh:51]: typo `etup`

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, please find some comments/questions along the code. Thanks

File: bin/iam-token.sh (hunk 1 above)
Comment (vkuznet) [reply bin/iam-token.sh:2]: done

File: bin/iam-token.sh (hunk 2 above)
Comment (vkuznet) [reply bin/iam-token.sh:6]: Alan, please state it either way, I can't interpret maybe. To me `IAM_TOKEN` is totally fine, if you want it to be renamed please write this explicitly without question marks. SO far I leave it as is

File: bin/iam-token.sh (hunk 4 above)
Comment (vkuznet) [reply bin/iam-token.sh:4]: I added it to the comments, these are values which we obtain from IAM provider, i.e. it is not file names.

File: bin/iam-token.sh (hunk 6 above)
Comment (vkuznet) [reply bin/iam-token.sh:51]: done

File: bin/iam-token.sh (hunk 5 above)
Comment (vkuznet) [reply bin/iam-token.sh:45]: I added corresponding comment and aligned the block

**Comment by amaltaro** [comment]
@vkuznet Valentin, this is looking good to me now, thanks for making those changes. Are you still working on it, if not please remove that label and fire up another review request.

To make sure we have these changes documented somewhere - other than GH comments - could you please populate this wiki with the necessary content: https://github.com/dmwm/WMCore/wiki/Tokens-in-WMCore

Feel free to change sections and etc, of course.

**Comment by amaltaro** [comment]
@vkuznet Valentin, this is looking good to me now, thanks for making those changes. Are you still working on it, if not please remove that label and fire up another review request.

To make sure we have these changes documented somewhere - other than GH comments - could you please populate this wiki with the necessary content: https://github.com/dmwm/WMCore/wiki/Tokens-in-WMCore

Feel free to change sections and etc, of course.

**Review by amaltaro — APPROVED** [review]
It looks good to me. Thanks

**Pull request merged**